### DISCUSSION TITLE
Unable to import module in setup.py (using setuptools and pyproject.toml)

### DISCUSSION THREAD

**Discussion by luclaurent** [pypa/setuptools#3868] [Q&A] [answered]
I am currently using setuptools to build and install a package using a `pyproject.toml` and a `setup.py` files. I currently try to import usefule package in `setup.py` in order to prepare the building of Fortran's extensions. But I have an error on importation of module. I am not sure that the trouble comes from `setuptools` but maybe you willa have an idea.

I currently have the following tree

    .
    ├── MANIFEST.in
    ├── README.md
    ├── AAA
    │   ├── __init__.py
    │   └── run.py
    ├── pyproject.toml
    ├── setup.py
    ├── setup_misc
    │   ├── __init__.py
    │   └── setup_misc.py

In the `setup.py` file, I import a package using

    from setup_misc import setup_misc as s
    from setuptools import setup
    from numpy.distutils.core import Extension

The purpose is to install package `AAA` but the `setup.py` requires functions that are declared in `setup_misc.py`

I have added the following line in `MANIFEST.in`

    recursive-include setup_misc *

and `pyproject.toml` contains

    [build-system]
    requires = ["setuptools","numpy","wheel"]
    build-backend = "setuptools.build_meta"

I run the setup using `pip install -v .` and it returns:
 

    ModuleNotFoundError: No module named 'setup_misc'

Is it possible to import module in `setup.py`? If yes, what is the right way?

**Comment by abravalheri** [comment, marked as answer]
Hi @luclaurent, `pip` uses PEP 517 hooks to call setuptools and no longer runs deprecated `python setup.py ...` command. This impacts on the way `sys.path` is processed by Python.

If you need the project directory to be added to `sys.path`, I recommend you explicitly doing that in your `setup.py` before the other imports. Something like:

```python
import os, sys
sys.path.insert(0, os.path.dirname(__file__))
...
```
(Please double check because I wrote this from the top of my head)

[^1]: When you run a script with `python script_name.py` Python will automatically prepend the script’s directory to `sys.path`. This allows you to import modules not installed in your system. See (https://docs.python.org/3/library/sys.html#sys.path)

**Reply by luclaurent** [reply]
It works fine. Thank you !!
